internal/eru/resources/core.go

```diff
@@ -0,0 +1,234 @@
+package resources
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/projecteru2/core/log"
+	cpumemtypes "github.com/projecteru2/core/resource/plugins/cpumem/types"
+	resourcetypes "github.com/projecteru2/core/resource/types"
+	stotypes "github.com/projecteru2/resource-storage/storage/types"
+	"github.com/projecteru2/yavirt/configs"
+	"github.com/projecteru2/yavirt/internal/eru/types"
+	"github.com/projecteru2/yavirt/internal/utils/notify/bison"
+	gputypes "github.com/yuyang0/resource-gpu/gpu/types"
+)
+
+// CoreResourcesManager used to cache the resources on core
+type CoreResourcesManager struct {
+	mu     sync.Mutex
+	cpumem *cpumemtypes.NodeResource
+	gpu    *gputypes.NodeResource
+	sto    *stotypes.NodeResource
+}
+
+func NewCoreResourcesManager() *CoreResourcesManager {
+	return &CoreResourcesManager{}
+}
+
+func (cm *CoreResourcesManager) fetchResources() {
+	ctx, cancel := context.WithTimeout(context.TODO(), 30*time.Second)
+	defer cancel()
+	logger := log.WithFunc("fetchResources")
+
+	resp, err := cli.GetNodeResource(ctx, configs.Hostname())
+	if err != nil {
+		logger.Errorf(ctx, err, "failed to fetch resource from core")
+		return
+	}
+	capacity := resp.Capacity
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	for resName, rawParams := range capacity {
+		switch resName {
+		case "cpumem":
+			cpumem := cpumemtypes.NodeResource{}
+			if err = mapstructure.Decode(rawParams, &cpumem); err != nil {
+				logger.Errorf(ctx, err, "failed to unmarshal resource cpumem")
+			} else {
+				logger.Debugf(ctx, "[core fetchResources] cpumem: %v", cpumem)
+				cm.cpumem = &cpumem
+			}
+		case "storage":
+			sto := stotypes.NodeResource{}
+			if err = mapstructure.Decode(rawParams, &sto); err != nil {
+				logger.Errorf(ctx, err, "failed to unmarshal resource storage")
+			} else {
+				logger.Debugf(ctx, "[core fetchResources] storage: %v", sto)
+				cm.sto = &sto
+			}
+		case "gpu":
+			gpu := gputypes.NodeResource{}
+			if err = mapstructure.Decode(rawParams, &gpu); err != nil {
+				logger.Errorf(ctx, err, "failed to unmarshal resource gpu")
+			} else {
+				logger.Debugf(ctx, "[core fetchResources] gpu: %v", gpu)
+				cm.gpu = &gpu
+			}
+		}
+	}
+}
+
+func (cm *CoreResourcesManager) GetCpumem() (ans *cpumemtypes.NodeResource) {
+	cm.mu.Lock()
+	ans = cm.cpumem
+	cm.mu.Unlock()
+	if ans != nil {
+		return
+	}
+	cm.fetchResources()
```

Comment:
rename to `fetchResourcesWithLock` or `FetchResources` (with lock implicitly)
i'd like to the former and 
```
cm.mu.Lock()
defer cm.mu.Unlock()
...
```
only once